lib/src/main/java/io/ably/lib/util/AgentHeaderCreator.java

```diff
@@ -0,0 +1,61 @@
+package io.ably.lib.util;
+
+import android.os.Build;
+import io.ably.lib.BuildConfig;
+import io.ably.lib.transport.Defaults;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public class AgentHeaderCreator {
+    /**
+     * Separates agent entries from each other.
+     */
+    private static final String AGENT_ENTRY_SEPARATOR = " ";
+
+    /**
+     * Separates agent name from agent version.
+     */
+    private static final String AGENT_DIVIDER = "/";
+    private static final String ANDROID_LIBRARY_NAME = "android";
+
+    public static String create(Map<String, String> additionalAgents) {
+        StringBuilder agentStringBuilder = new StringBuilder();
+        if (!additionalAgents.isEmpty()) {
+            agentStringBuilder.append(getAdditionalAgentEntries(additionalAgents));
+        }
+        agentStringBuilder.append(Defaults.ABLY_AGENT_VERSION);
+        if (BuildConfig.LIBRARY_NAME == ANDROID_LIBRARY_NAME) {
+            agentStringBuilder.append(AGENT_ENTRY_SEPARATOR);
+            agentStringBuilder.append(getAndroidAgent());
+        }
+        return agentStringBuilder.toString();
+    }
+
+    private static String getAdditionalAgentEntries(Map<String, String> additionalAgents) {
+        StringBuilder additionalAgentsBuilder = new StringBuilder();
+        for (String additionalAgentName : getSortedAgentNames(additionalAgents)) {
+            String additionalAgentVersion = additionalAgents.get(additionalAgentName);
+            additionalAgentsBuilder.append(additionalAgentName);
+            if (additionalAgentVersion != null) {
+                additionalAgentsBuilder.append(AGENT_DIVIDER);
+                additionalAgentsBuilder.append(additionalAgentVersion);
+            }
+            additionalAgentsBuilder.append(AGENT_ENTRY_SEPARATOR);
+        }
+        return additionalAgentsBuilder.toString();
+    }
+
+
+    private static List<String> getSortedAgentNames(Map<String, String> agents) {
```

Comment:
Nice, but there's no need to sort them.

If you're sorting them to make testing easier then I would handle that test-side.

Reply:
The precise origin of this comes from @paddybyers comment [here](https://github.com/ably/docs/pull/1034#discussion_r604345679), although it wasn't subsequently discussed in the thread. I do basically agree with it though; I think we should recommend an ordering just because it's nice to be able to say "this is precisely what the ably-agent header should look like" rather than having a bunch of equally correct possibilities. To me, dependency order makes the most sense semantically but I do also realise it might be ambiguous at times. I wouldn't personally be against changing the recommendation to something like alphabetical order just so that it's totally unambiguous for every implementation.

Reply:
I think I'm ok with the idea of the library codebase getting pride of place (i.e. `ably-java`), but I'm not sure I see the value in any ordering beyond that... especially when it can be ambiguous as you say. Alphabetical, perhaps, but what value would that really add? There's got to be some concrete information we convey to the receiver (Ably's service) by doing this, as otherwise I don't really get the value in ordering at all.

Reply:
There is an expectation that the Ably SDK version comes first in the list, and we're relying on that in the router to maintain the current metrics that power the [client library statistics dashboard](https://grafana.ably.io:8443/d/000000029/client-library-statistics?orgId=1) by emitting a metric for the first entry in the list, as it was considered unacceptable to break that dashboard.

There has been [discussion in Slack](https://ably-real-time.slack.com/archives/CURL4U2FP/p1621854952027600) about how we expose the rest of the entries in the list for later querying, since we aren't comfortable with just turning every entry into its own metric. The current plan is for the router to log the header which can then be queried in Athena, but there has been some debate on whether that is the correct approach.

Reply:
I've removed the sorting https://github.com/ably/ably-java/pull/671/commits/522f29be7c671272604b4014d8c7d51002dc5e73 and placed the library agent as the first one in the agent list https://github.com/ably/ably-java/pull/671/commits/22258979d0770976c06c9e689d193214b817c78b

Reply:
Thanks. @owenpearson is going to lead in terms of bringing this conversation to a conclusive close.